### PR TITLE
perf(nostr): verify inbound relay events off the main isolate

### LINKED ISSUE #5863: perf(nostr): relay event signature verification runs on the main isolate (app-wide jank)
## Summary

App-wide lag reported on Discord (profile counts stuck on "—" for ~4s at cold
start; slow, choppy camera preview that smooths out once recording starts) traces
to **relay event ingestion doing cryptographic signature verification on the main
(UI) isolate** for every incoming event — ungated, and **deduplicated *after*
verify**, so the same event arriving from N relays is verified N times.

The team's working theory was "background DM decryption starves the app." That is
**misattributed**: DM crypto is already off the main isolate
(`DmVerifyIsolate` / `DmDecryptWorker`). The real continuous main-isolate hog is
relay-event verify.

## Evidence (main)

- Path (all on the main isolate): `relay_base.dart:105-115` `jsonDecode` →
  `relay_pool.dart:151` `_onEvent` → `:416` `Event.fromJson` → `:417`
  `!event.isValid || !event.isSigned`.
- Cost: `event.dart:130` sha256 id recompute + `event.dart:141` bip340
  `schnorr.verify`. **Benchmarked ~6.5 ms per verify** (AOT) — 99.6% of per-event
  cost; sha256-id and jsonDecode are ~14 µs each.
- **Dedup happens AFTER verify** (`relay_pool.dart:473`), so the same event id from
  each connected relay (divine + indexers + NIP-65 user relays, ~4–8) pays full
  verify; reconnect/`autoSubscribe` replays re-verify too.
- The EVENT branch (`relay_pool.dart:408-476`) has **no `await`** before dispatch,
  so a cold-start backfill burst blocks the event loop (~50 events × 6.5 ms ≈
  325 ms/relay on a Mac; more on ARM), starving the profile-count REST
  continuation and frame callbacks.
- The camera preview is a native GPU `Texture`
  (`divine_camera/.../camera_preview_widget.dart:249`), but the **UI thread** must
  composite each frame — the same thread verify monopolizes → choppy preview that
  smooths once recording engages `CodecHeavySurfaceCubit`.
- Off-main infra already exists but is unused for feed relays: `EventSignChecker`
  worker-pool is dead code; `RelayIsolate` (off-main verify) is wired only for the
  NIP-46 signer; feed/indexer relays are all `RelayBase` (main-isolate verify).

## Proposed fixes (ranked)

1. **Dedup-before-verify — do first (Low effort, ~50–75% cold-start reduction).**
   Bounded LRU seen-id set at the top of `_onEvent`'s EVENT branch, checked before
   `Event.fromJson`/`isValid`/`isSigned`; on hit, skip the crypto (still accumulate
   `event.sources`). Insert into the set only after verify passes. Kills the
   N-relay fan-out multiplier. Localized to `nostr_sdk` `relay_pool.dart`,
   unit-testable.
2. **Move verify off the main isolate (Medium — biggest absolute relief).** Stand
   up a persistent key-less verify isolate (reuse the `DmVerifyIsolate` pattern);
   batch-verify decoded EVENT frames between `relay_base` decode and `_onEvent`
   routing; preserve per-subscription ordering; add back-pressure. Removes the
   ~6.5 ms/event from the UI thread, including for unique events dedup can't help.
3. **Device-tier throttle (Low–Medium).** Wire the existing `DeviceMemoryUtil` to
   pace verify batch size / isolate concurrency on `MemoryTier.low`.
4. **Cold-start-specific — decouple profile counts from `nostrReady`.** The
   profile-count fetch is a public REST call needing no signer, but it's gated
   behind relay-connect + NIP-42 AUTH (`repository_providers.dart:285-291`,
   `profile_stats_row_widget.dart:35-37`). Gate it on **identity-known (pubkey
   available)** instead so counts render as soon as REST returns — directly attacks
   the "—" symptom. Also defer DM gift-wrap + discovery-feed subscriptions past
   first frame and narrow the DM `since` (2-day) backlog.
5. **Camera nit.** Drop the redundant second `_disposeVideoControllers()` in
   `_initializeCamera` (`video_recorder_screen.dart:172`, already run at :293);
   optionally extend the codec-heavy guard to throttle ingest during preview.

**Not recommended by default:** trust-gating the first-party relay to skip verify
(`relay.divine.video` validates server-side) — removes client-side tamper
detection; security trade-off, keep behind an explicit decision.

## Acceptance criteria

- Cold-start profile counts render promptly (no ~4s "—").
- Camera preview is smooth from open (no choppy warm-up).
- Signature verify is deduped-before-verify and/or off the main isolate; measured
  improvement on a low-end device (the reporter's class of device).

## Notes

Sizeable — a good split is: (a) dedup-before-verify quick win, (b) off-main verify
isolate, (c) decouple profile-count fetch, (d) camera/cold-start nits. Evidence
gathered during the #5854 investigation (perf deep-dive).

### PR DESCRIPTION
**Pull request by realmeylisdev**
Closes #5863

## What & why

Runs the inbound relay-event **Schnorr signature verify off the main isolate**, composed on top of #5888 (now merged). #5888 added the `(id, sig)` dedup — session-dup and known-verified (seeded from the local store) events skip verify entirely. This PR moves the *remaining* fresh verify — the events #5888 does **not** already trust — off the main/UI isolate, so the cold-start burst of genuinely-new events (and steady-state new events) no longer blocks frame callbacks.

## What's in this PR

- **`EventVerifyIsolate`** — a keyless, long-lived background isolate that recomputes the event id + verifies the Schnorr signature. Nothing secret crosses the port (verification needs no private key). Mirrors `dm_repository`'s `DmVerifyIsolate`, but persistent rather than drain-scoped.
- Slots into #5888's verify ladder: `_verifySignatureOffMain` replaces the `else if (event.isSigned)` fresh-verify branch, so only events that fall through the session-dup / known-verified skips reach the worker.
- Wired **opt-in** via `NostrClientConfig.eventVerifyWorkerSpawner`, spawned in `initialize()` right after #5888's `_seedVerifiedEventIds()` and before relays connect; the app passes `EventVerifyIsolate.spawn` (null on web/tests → inline main-isolate verify, unchanged).
- EVENT/EOSE frames are **serialized per relay** so the now-async verify preserves in-order delivery (an EOSE must not complete a query before its preceding events finish verifying).
- **Inline fallback** on any worker failure, so a verify never hangs. **Dispatch is unchanged** — every verified event flows through the normal path, so per-relay `noteReceive()` / `event.sources.add(relay.url)` source attribution is fully preserved.

## Testing

- `event_verify_isolate_test.dart` — spawn/verify/close, malformed payload → `false`, post-close throws.
- `relay_pool_offmain_verify_test.dart` — worker verify routing, worker-reject drop, inline fallback on throw, per-relay in-order delivery under out-of-order verify.
- #5888's `relay_pool_verify_dedup_test.dart` stays green (the worker-less path is unchanged), plus the `nostr_client` initialize/dispose suites.

### CODE REVIEW

**Review by hm21 — COMMENTED** [review]
Heads-up: we've already been working on exactly this in **#5888** — dedup relay events before the schnorr verify in `relay_pool.dart`. Rather than two parallel implementations, let's converge on one. Sharing our on-device profiling so it's clear where the win actually comes from:

**Cold-start relay signature verify:**

| approach | relay-path verify | `Event.isSigned` (total) |
|---|---|---|
| baseline | ~38% of startup CPU | 42.5% |
| within-session / cross-relay dedup by id (≈ part 1 here) | ~36% | **40.6%** (−2pp) |
| **#5888** — seed already-verified `(id, sig)` from the local store | **~1%** | **2.7%** |

(last row confirmed in a profile-mode / AOT run on device.)

The dominant cost is **cross-session re-sends** — events already in the local store that relays replay on *every* cold start. Part 1's LRU starts empty each launch, so it can't skip those; it only catches cross-relay duplicates *within* a session, which our numbers show is the small slice (~2pp).

So the ask: build on top of #5888 instead of parallel. Two things from here are great additions to fold in:
- your **drop-before-dispatch** (skip `Event.fromJson` + dispatch, not just the verify) — a nice optimization on top of the cross-session seed;
- **part 2** (decoupling profile counts from `nostrReady`) is a clean standalone win, independent of all this.

Happy to help rebase your changes onto #5888. nit: dropping duplicates before dispatch skips `event.sources.add(relay.url)` — check nothing downstream needs per-relay source tracking.

**Comment by realmeylisdev** [comment]
Thanks for the on-device numbers, @hm21 — that settles it. The empty-start LRU only catches within-session cross-relay dupes, which your table shows is the ~2pp slice; the real cost is the cross-session re-sends that your `(id, sig)` seed-from-store in #5888 eliminates (42.5% → 2.7%). Agreed — no parallel implementations, let's converge on #5888.

Concretely:

- **Profile-count decouple (part 2)** is now out of this PR and standing on its own in **#5901** — it's independent of all the verify work, green, and shouldn't wait on the dedup convergence.
- **Off-main verify (`EventVerifyIsolate`)** + **drop-before-dispatch** I'll rebase onto #5888 once it's landed/green, dropping my empty-start LRU in favor of your seeded set. The off-main isolate is orthogonal to *which* dedup wins — it takes the ~6.5 ms/event schnorr verify off the UI thread even for the distinct events the seed can't skip — so it's purely additive on top of #5888.

On the `event.sources` nit — you're right, and it's load-bearing. My drop path `return`s before `Event.fromJson`, so a cross-relay duplicate never reaches `event.sources.add(relay.url)`, and both `EventMemBox`'s source merge (`event_mem_box.dart:126-160`) and the nip95 relay-hint encoding consume that set. So drop-before-dispatch has to still record the source relay on a dup hit rather than bail early. I'll fold it in that way when I rebase onto #5888.

I'll take you up on the rebase help — I'll ping here once #5888 is in.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 1]
```diff
@@ -413,15 +530,42 @@ class RelayPool {
...
+        // no verify, and no dispatch (downstream consumers dedup by id anyway).
+        final rawId = eventJson['id'];
+        if (rawId is String && _verifiedSeenIds.containsKey(rawId)) {
+          // Touch as most-recently-seen (LRU) and drop the duplicate.
+          _verifiedSeenIds
+            ..remove(rawId)
+            ..[rawId] = true;
+          _duplicateEventCount++;
+          _maybeLogDedupStats();
+          return;
```
Comment (NotThatKindOfDrLiz) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: This early return drops a duplicate before `event.sources.add(relay.url)` and before dispatch, so downstream source merging never learns that this relay also served the event. That is load-bearing today: `EventMemBox` merges source relays from duplicate dispatched events, `VideoEvent` falls back to `event.sources` for `sourceRelay`, and NIP95 `nevent` relay hints use `result.sources`. The duplicate path needs to preserve per-relay source attribution before it can safely skip dispatch.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 2]
```diff
@@ -383,6 +436,51 @@ class RelayPool {
     return null;
   }
 
+  /// #5863 instrumentation (TEMPORARY — remove after the on-device
```
Comment (NotThatKindOfDrLiz) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: This temporary measurement block should be removed before merge. It is explicitly labeled `TEMPORARY`, logs profile/debug-only measurement data from the hot relay path, and includes an emoji log message. The `@visibleForTesting` counters that back tests can stay, but this instrumentation should not ship in the PR.

**Review by NotThatKindOfDrLiz — DISMISSED** [review]
Requesting changes rather than takeover here. This PR should not merge in its current shape: the dedup implementation is now explicitly superseded by #5888, the profile-count split is already standing alone in #5901, and the remaining off-main verify work needs to be rebased onto the winning dedup path. Please drop the temporary LRU/instrumentation from this PR, let #5901 carry the profile-count decouple, and reintroduce EventVerifyIsolate/drop-before-dispatch on top of #5888 with duplicate-hit source attribution preserved.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 3]
```diff
@@ -413,15 +530,42 @@ class RelayPool {
         final eventJson = _mapAt(relay, json, 2, 'EVENT payload');
         if (eventJson == null) return;
 
+        // Dedup BEFORE the expensive signature verify (#5863). The same event
+        // id is delivered once per connected relay, and the schnorr verify
+        // below is the dominant per-event cost on the main isolate. Read the id
+        // straight from the raw payload — a duplicate needs no Event.fromJson,
+        // no verify, and no dispatch (downstream consumers dedup by id anyway).
+        final rawId = eventJson['id'];
+        if (rawId is String && _verifiedSeenIds.containsKey(rawId)) {
```
Comment (rabble) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: This duplicate fast path returns before the normal receive path records relay metadata. A second relay delivering the same verified event now skips `relay.relayStatus.noteReceive()` and never reaches `event.sources.add(relay.url)` below. Current downstream code relies on duplicate delivery reaching merge logic so relay sources are preserved; otherwise `sourceRelay`/diagnostics depend on whichever relay won the race. Please skip only the expensive verify/dispatch work while still preserving duplicate-hit relay status/source metadata, and add a test that both relay sources survive a duplicate delivery.

**Review by rabble — DISMISSED** [review]
Change requested: the duplicate-before-verify fast path drops relay receive/source metadata for duplicates. I left an inline comment on relay_pool.dart.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart (hunk 1 above)
Comment (realmeylisdev) [reply mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: Fixed in 3f30adeb58 by removing the duplicate early-return entirely — the dedup-before-verify path moves to #5888. This PR now only moves the Schnorr verify off the main isolate and never skips dispatch, so `event.sources.add(relay.url)`, `EventMemBox` source merging, `VideoEvent.sourceRelay`, and the NIP-95 relay hints all run for every delivery.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart (hunk 2 above)
Comment (realmeylisdev) [reply mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: Removed in 3f30adeb58 along with the rest of the dedup path — `_maybeLogDedupStats`, the LRU, and the `@visibleForTesting` counters are all gone. The off-main verify that remains has no hot-path logging.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart (hunk 3 above)
Comment (realmeylisdev) [reply mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: Fixed in 3f30adeb58 — the duplicate fast path is removed, so a duplicate delivery now goes through the unchanged normal dispatch path: every delivery still hits `relay.relayStatus.noteReceive()` and `event.sources.add(relay.url)`, so both relays' sources survive (same as `main` today). Dedup moves to #5888, which keeps full dispatch. The "both sources survive a duplicate" test belongs with #5888's fast path rather than here, since this PR no longer has one — the only change here is running the verify off the main isolate.

**Comment by realmeylisdev** [comment]
Restructured to the review consensus in 3f30adeb58:

- **Dropped** the dedup-before-verify LRU + temporary instrumentation — converging on @hm21's #5888, which dedups better (`(id, sig)`-keyed, cross-session seed from the local store) and keeps full dispatch.
- **Dropped** the profile-count decouple — already shipped as #5901.
- **Kept** only the off-main `EventVerifyIsolate` verify. It dispatches every event through the normal path, so per-relay source/status attribution is preserved and the inline blockers are resolved by removal (no fast path left to skip `noteReceive()` / `sources.add`).

Left as a **draft** and held until #5888 merges: it will then rebase onto `main` so the isolate composes on top of the `(id, sig)` skip (only genuinely-new events reach the worker) — the "on top of #5888" @NotThatKindOfDrLiz asked for, without stacking PRs. cc @rabble

**Comment by realmeylisdev** [comment]
Rebased onto `main` now that #5888 has merged, and composed the off-main verify on top of it: `_verifySignatureOffMain` replaces #5888's `else if (event.isSigned)` fresh-verify branch, so only events that fall through the session-dup / known-verified skips reach the isolate. Dispatch is unchanged, so per-relay `noteReceive()` / `sources.add` attribution stays intact (the earlier blocker). Green locally: nostr_sdk 52 tests (incl. #5888's `relay_pool_verify_dedup_test`) + nostr_client 153 + `dart analyze`.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Re-reviewed current head 19a3fdd after the restructure. The stale blockers targeted removed fast-path/instrumentation code; the remaining off-main verify path preserves dispatch/source attribution and has green CI. Approving.
